src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfReader.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2016 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.ac.ebi.eva.pipeline.io.readers;
+
+import htsjdk.variant.vcf.VCFCompoundHeaderLine;
+import htsjdk.variant.vcf.VCFConstants;
+import htsjdk.variant.vcf.VCFContigHeaderLine;
+import htsjdk.variant.vcf.VCFFileReader;
+import htsjdk.variant.vcf.VCFFilterHeaderLine;
+import htsjdk.variant.vcf.VCFFormatHeaderLine;
+import htsjdk.variant.vcf.VCFHeader;
+import htsjdk.variant.vcf.VCFHeaderLine;
+import htsjdk.variant.vcf.VCFInfoHeaderLine;
+import htsjdk.variant.vcf.VCFSimpleHeaderLine;
+import org.opencb.biodata.formats.variant.vcf4.VcfAlternateHeader;
+import org.opencb.biodata.formats.variant.vcf4.VcfFilterHeader;
+import org.opencb.biodata.formats.variant.vcf4.VcfFormatHeader;
+import org.opencb.biodata.formats.variant.vcf4.VcfInfoHeader;
+import org.opencb.biodata.models.variant.Variant;
+import org.opencb.biodata.models.variant.VariantSource;
+import org.springframework.batch.item.file.FlatFileItemReader;
+import org.springframework.core.io.FileSystemResource;
+import org.springframework.core.io.Resource;
+
+import uk.ac.ebi.eva.pipeline.io.GzipLazyResource;
+import uk.ac.ebi.eva.pipeline.io.mappers.VcfLineMapper;
+import uk.ac.ebi.eva.utils.CompressionHelper;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * VCF file reader.
+ * <p>
+ * This Reader uses a {@link VcfLineMapper} to parse each line, and {@link VCFFileReader} from htslib to fill the
+ * VariantSource with metadata from the VCF header.
+ * <p>
+ * It doesn't matter if the file is compressed or not.
+ */
+public class VcfReader extends FlatFileItemReader<List<Variant>> {
+    /**
+     * You can retrieve the header of the VCF like this: `source.getMetadata().get(VARIANT_FILE_HEADER_KEY)`,
+     * but why "variantFileHeader"? we are using the converter
+     * {@link org.opencb.opencga.storage.mongodb.variant.DBObjectToVariantSourceConverter}, which requires it, and will
+     * store it in mongo as "header".
+     */
+    public static final String VARIANT_FILE_HEADER_KEY = "variantFileHeader";
+
+    private final VariantSource source;
```

Comment:
What will happen if the step fails while processing a variant? This object will be lost, right?

This could be implemented in another reader for header entries only, that will stop reading once the first line without a # is reached. And the variant reader would mark as comments the lines prefixed with #.